Partial Fix for Memory issue - Limiting the number of ImageSources cached

alekhyareddy28 opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
* The memory used by PT Run was increasing on consecutive launches. One of the reasons for this to happen was that all the image icons were cached. 
* Even though the code had an upperlimit on the `maxCached` values which was set to 5000, this part of the code was never called unless powerToys was closed. 
* Therefore, since all the icons were being cached, the size of the dictionary continued to increase.

**Work around**-
* Instead of caching all images and increasing the size of the dictionary, the following changes have been made -
1. Reduce the number of cached images to 50.
2. If the number of images increase over the permissible factor (set to 2) of the cached images (ie. 2 * 50), then the ones which are infrequently used would be removed from the dictionary and there be 50 images that are cached.
This ensures that the dictionary size is always less than 100.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #2047.
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx


<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->

## Validation Steps Performed

The changes in this PR can be validated by following these steps -
1. Take snapshots of the memory usage by using the Diagnostic tools of Visual studio (Debug -> Windows -> Diagnostic tools).
2. Take the first snapshot after all the dlls are loaded and the memory utilization is stabilized.
3. Take the next memory snapshot after many queries (don't have to execute an application). The aim is to just load many icons.
4. Compare the second snapshot by setting the first one as the baseline.
5. Search for memory size diff in bytes for Concurrent Dictionary with value ImageSource.
6. Compare this value to that obtained without the changes in this PR, to get the results as shown in the following images.

* Before this change (after about twenty searches)-
![image](https://user-images.githubusercontent.com/28739210/85344943-e3a3fc80-b4a5-11ea-919e-6226a3132b16.png)

* After this change (after about twenty searches) -
![image](https://user-images.githubusercontent.com/28739210/85344930-df77df00-b4a5-11ea-98bf-cc5fc58c9c72.png)


## Outstanding tasks
* Even though this fixes the issue partially, the memory consumed by PT Run is still high. It starts off with 160/170 MB and after first launch goes up to 230/250 MB, following which it increases a little on each run.
* Another area where the memory utilization can be improved is the `Dictionary<int64, WeakReference>`. The following snapshots are after searching for a few apps using PT Run -
![image](https://user-images.githubusercontent.com/28739210/85345097-63ca6200-b4a6-11ea-80b3-fc80a78c9841.png)
![image](https://user-images.githubusercontent.com/28739210/85345103-69c04300-b4a6-11ea-899b-99bbaaa11e07.png)

It seems like this is related to the `ResourceDictionary` of the UI components. The number of elements in this dictionary and the components that they are referring to seem to increase on each run. This needs further investigation.

## Additional Information
* I tried to set a hard limit on the heap size using `runtime.config.json` and `System.GC.HeapHardLimit`. I did not notice any drastic differences in the Garbage collection when the limit was set to values ranging from 20 MB to 200 MB. Also, this might differ from system to system so I did not go down this route.

References - https://docs.microsoft.com/en-us/dotnet/core/run-time-config/garbage-collector

arjunbalgovind commented on src/modules/launcher/Wox.Infrastructure/Image/ImageCache.cs:
```diff
@@ -22,7 +23,30 @@ public class ImageCache
+
+                // To prevent the dictionary from drastically increasing in size by caching images, the dictionary size is not allowed to grow more than the permissibleFactor * maxCached size
+                // This is done so that we don't constantly perform this resizing operation and also maintain the image cache size at the same time
+                if (_data.Count > permissibleFactor * MaxCached)
```
Can you clarify the difference between the cache and the dictionary? Is cache referring to what is stored in a file versus dictionary referring to what is stored in memory?

alekhyareddy28 replied:
When launcher is running, the information about images is stored in the dictionary with the path as the key and the Bitmap Imagesource as the value. So the dictionary is in memory (heap) and it continues to grow whenever the user searches for apps/files with unique icons (a hash was calculated using each image to prevent storing duplicates). When the app is closed, the information present in the dictionary would be stored in an Image cache file (on the disk) up to 5000 images which would again be loaded into the dicitonary (memory) at startup. As image Sources are being stored in the dictionary, it is memory intensive compared to storing only integers or strings. Hope this clarifies the difference.

arjunbalgovind commented:
I launched it from the VS debugger and after hitting Alt+Space mem usage was at 150MB. I did a search query for `a` and I just scrolled all the way till the bottom and the mem footprint went to 340 MB after which I would see a sudden drop (maybe the Cleanup step?). I did notice though that when the GC kicks in, the UI freezes a bit. On my first run I was able to reach 450MB as well and I also hit an exception in the Cleanup method. Not sure if this is a handled exception though.
![image](https://user-images.githubusercontent.com/32061677/85449067-2f10e600-b54c-11ea-8b16-5560137ca17f.png)

alekhyareddy28 commented:
> I launched it from the VS debugger and after hitting Alt+Space mem usage was at 150MB. I did a search query for `a` and I just scrolled all the way till the bottom and the mem footprint went to 340 MB after which I would see a sudden drop (maybe the Cleanup step?). I did notice though that when the GC kicks in, the UI freezes a bit. On my first run I was able to reach 450MB as well and I also hit an exception in the Cleanup method. Not sure if this is a handled exception though.
> ![image](https://user-images.githubusercontent.com/32061677/85449067-2f10e600-b54c-11ea-8b16-5560137ca17f.png)

I tried to repro the exception and wasn't able to but it seems like that happens when we use OrderBy with concurrent dictionary on multiple threads. I removed Task.Run and instead am doing it on the same thread. This should not throw any exceptions. I profiled the time for reordering the dictionary and it seems to take upto a maximum of 10 ms, so I think it should be fine performing the operation on the same thread.
Please let me know if the exception still persists or if you notice any visible performance issue.

Regarding GC, that is not something that I have modified. It is the default .NET GC and it might be something that we can look into later. GC should be in a background thread and technically not cause any stutter.

Regarding the 150 MB - 340 MB increase, a major portion of that increase is due to everything being indexed during  that time. Try to execute the query after the memory has stabilised. You should not see any drastic drop. 
![image](https://user-images.githubusercontent.com/28739210/85458232-bf9ff400-b555-11ea-90f3-57151eb43fa4.png)
In the above image, I have not executed anything. 

If I put a lot of pressure on the system, this is how it appears for me - 
![image](https://user-images.githubusercontent.com/28739210/85458443-f83fcd80-b555-11ea-8c98-01c016472bca.png)
NOTE: GC would not take place this often usually.

arjunbalgovind commented on src/modules/launcher/Wox.Infrastructure/Image/ImageLoader.cs:
```diff
@@ -184,9 +184,13 @@ public static ImageSource Load(string path, bool loadFullImage = false)
                     if (GuidToKey.TryGetValue(hash, out string key))
                     { // image already exists
-                        img = ImageCache[key];
+                        if(ImageCache.Usage.TryGetValue(path, out ImageCacheValue))
```
[nit] Earlier this was only checking whether GuidToKey contained the hash and if so it would retrieve the image. Why is a change required here for checking the ImageCache as well?

alekhyareddy28 replied:
Due to resizing of the dictionary, there are cases when the key is not a part of the imageCache. That's why this has been added to check if the img exists even after the resizing. Without this condition, it throws an exception while trying to retrieve that item.

arjunbalgovind commented on src/modules/launcher/Wox.Infrastructure/Image/ImageCache.cs:
```diff
@@ -2,17 +2,18 @@
         public ConcurrentDictionary<string, int> Usage = new ConcurrentDictionary<string, int>();
         private readonly ConcurrentDictionary<string, ImageSource> _data = new ConcurrentDictionary<string, ImageSource>();
-
+        private const int permissibleFactor = 2;
```
[nit] Any reason for using `2`?

alekhyareddy28 replied:
I was going along the lines of vector resizing. The size increments/decrements are always factors of two. No other reason.

arjunbalgovind commented on src/modules/launcher/Wox.Infrastructure/Image/ImageCache.cs:
```diff
@@ -22,7 +23,26 @@ public class ImageCache
+                        int dictValue;
+                        if (!Usage.TryGetValue(key, out dictValue))
+                        {
+                            ImageSource test;
```
[nit] variable name

arjunbalgovind commented on src/modules/launcher/Wox.Infrastructure/Image/ImageCache.cs:
```diff
@@ -22,7 +23,26 @@ public class ImageCache
+                if (_data.Count > permissibleFactor * MaxCached)
+                {
+                    Cleanup();
+                    foreach (var key in _data.Keys)
```
[nit] Maybe add a comment to explain this part so that anyone reading this part doesn't need to go through Cleanup's definition to understand why this loop removes any entries that are not in the top used `maxCached` images.

somil55 commented on src/modules/launcher/Wox.Infrastructure/Image/ImageCache.cs:
```diff
@@ -22,7 +23,26 @@ public class ImageCache
+
+                // To prevent the dictionary from drastically increasing in size by caching images, the dictionary size is not allowed to grow more than the permissibleFactor * maxCached size
+                // This is done so that we don't constantly perform this resizing operation and also maintain the image cache size at the same time
+                if (_data.Count > permissibleFactor * MaxCached)
```
[clarification] How frequently do we hit `MaxCached` this limit? We are doing a clean up each time this happens, which seems to be a costly operation.

alekhyareddy28 replied:
As mentioned in my previous comment to Arjun, the maximum time it takes to do this is 10 ms (because the size of the dictionary is small). On trying it out on my machine, I did not see any performance issue. Maybe you can try it out and let me know if I should increase it to maybe 200. Right now it’s hit on every 100 unique image icons and it’s not very often. Cannot quantify a value because it depends on what the user searches for.

somil55 replied:
I didn't notice any visible difference but I think we should be able to support caching more images. Current code limits image size to 64*64 pixel bitmap image (~ 20KB) , so even with caching 1000 images we will never cross 20MB threshold.

alekhyareddy28 replied:
It's about preference. Do we want to cache 1000 images and use 20 MB of memory when we can rather cache just 50 images without any significant performance hit? I just tried to profile the reordering of the dictionary and it took around 24 ms with maxCached as 1000 on my system in comparison to 10 ms for 50.Feel free to correct me here but I think it's not about it being under 20 MB, rather about why we want to use 20 MB when we can get away with a few hundred KB. Let me know what you think. 

@ryanbodrug-microsoft, @crutkas do you have any preferences over the number of images to cache? Right now it is capped at 100, so whenever the images cached increase over 100, it is resized to 50 (trying to emulate the behavior of a vector).

alekhyareddy28 commented:
checking it in as is for now as it does not cause any regression but saves up on memory. Shall create an issue to profile the performance improvements that we get from using ImageCache.

alekhyareddy28 commented:
Created an issue to profile image loading and measure perf improvements on using ImageCache - https://github.com/microsoft/PowerToys/issues/4525.
